presto-parquet/src/main/java/com/facebook/presto/parquet/ColumnReaderFactory.java

```diff
@@ -40,43 +41,67 @@
 import com.facebook.presto.parquet.reader.ShortDecimalColumnReader;
 import com.facebook.presto.parquet.reader.TimestampColumnReader;
 import com.facebook.presto.spi.PrestoException;
+import org.apache.parquet.schema.LogicalTypeAnnotation.DecimalLogicalTypeAnnotation;
 
 import java.util.Optional;
 
-import static com.facebook.presto.parquet.ParquetTypeUtils.createDecimalType;
+import static com.facebook.presto.parquet.ParquetTypeUtils.isDecimalType;
+import static com.facebook.presto.parquet.ParquetTypeUtils.isShortDecimalType;
 import static com.facebook.presto.parquet.ParquetTypeUtils.isTimeStampMicrosType;
 import static com.facebook.presto.spi.StandardErrorCode.NOT_SUPPORTED;
-import static org.apache.parquet.schema.OriginalType.DECIMAL;
 import static org.apache.parquet.schema.OriginalType.TIMESTAMP_MICROS;
 import static org.apache.parquet.schema.OriginalType.TIME_MICROS;
 
 public class ColumnReaderFactory
 {
+    private static final Logger log = Logger.get(ColumnReaderFactory.class);
     private ColumnReaderFactory()
     {
     }
 
     public static ColumnReader createReader(RichColumnDescriptor descriptor, boolean batchReadEnabled)
     {
-        // decimal is not supported in batch readers
-        if (batchReadEnabled && descriptor.getPrimitiveType().getOriginalType() != DECIMAL) {
+        if (batchReadEnabled) {
             final boolean isNested = descriptor.getPath().length > 1;
             switch (descriptor.getPrimitiveType().getPrimitiveTypeName()) {
                 case BOOLEAN:
                     return isNested ? new BooleanNestedBatchReader(descriptor) : new BooleanFlatBatchReader(descriptor);
                 case INT32:
+                    if (!isNested && isShortDecimalType(descriptor)) {
+                        return new ShortDecimalFlatBatchReader(descriptor);
+                    }
                 case FLOAT:
                     return isNested ? new Int32NestedBatchReader(descriptor) : new Int32FlatBatchReader(descriptor);
                 case INT64:
                     if (isTimeStampMicrosType(descriptor)) {
                         return isNested ? new Int64TimestampMicrosNestedBatchReader(descriptor) : new Int64TimestampMicrosFlatBatchReader(descriptor);
                     }
+
+                    if (!isNested && isShortDecimalType(descriptor)) {
+                        int precision = ((DecimalLogicalTypeAnnotation) descriptor.getPrimitiveType().getLogicalTypeAnnotation()).getPrecision();
+                        if (precision < 10) {
+                            log.warn("PrimitiveTypeName is INT64 but precision is less then 10.");
```

Comment:
delete the warning; it's not actionable

or if I'm wrong and this is a real problem, then a waring isn't enough. This should fail outright